resources/testharness.js

```diff
@@ -1621,29 +1632,99 @@ policies and contribution forms [3].
      */
     Test.prototype.cleanup = function() {
         var error_count = 0;
-        var total;
+        var bad_value_count = 0;
+        function on_error() {
+            error_count += 1;
+            // Set test phase immediately so that tests declared within
+            // subsequent cleanup functions are not run.
+            tests.phase = tests.phases.ABORTED;
+        }
+        var this_obj = this;
+        var results = [];
+
+        this.phase = this.phases.CLEANING;
 
         forEach(this.cleanup_callbacks,
                 function(cleanup_callback) {
+                    var result;
+
                     try {
-                        cleanup_callback();
+                        result = cleanup_callback();
                     } catch (e) {
-                        // Set test phase immediately so that tests declared
-                        // within subsequent cleanup functions are not run.
-                        tests.phase = tests.phases.ABORTED;
-                        error_count += 1;
+                        on_error();
+                        return;
                     }
+
+                    results.push(result);
                 });
 
-        if (error_count > 0) {
-            total = this._user_defined_cleanup_count;
+        if (!this._is_promise_test) {
+            forEach(results,
+                    function(result) {
+                        if (result !== undefined) {
+                            bad_value_count += 1;
+                        }
+                    });
+            cleanup_done(this_obj, error_count, bad_value_count);
+        } else {
+            all_async(results,
+                      function(result, done) {
+                          if (result && typeof result.then === "function") {
+                              result
+                                  .then(null, on_error)
```

Comment:
Newbie question, what's the difference between this and `.then(done, on_error)`?

Reply:
For the chain `x().then(y, z)`, a roughly equivalent synchronous version would
look like this:

```js
var xThrew = false;
try {
  x();
} catch (error) {
  xThrew = true;
  z(error);
}

if (!xThrew) {
  y();
}
```

Compare that to `x().then(null, z).then(y)`:

```js
try {
  x();
} catch (error) {
  z(error);
}

y();
```

Reply:
Newbie indeed, some playing around in DevTools confirms the difference, the basic point I was missing is that handling the rejection causes the promise to be resolve with the return value of `on_error`, which is undefined.